#  4.8: Grouping Data & Aggregating Variables

## Table of content:

### Task 1 - Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

### Task 2 - In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

### Task 3 - Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

### Task 4 - Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

### Task 5 - The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

### Task 6 - The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:

#### 1) If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
#### 2) If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”


### Task 7 - In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

#### 1) If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
#### 2) If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
#### 3) If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

### Task 8 - Ensure your notebook is clean and structured and that your code is well commented.

### Task 9 - Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

#### Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

## Task 1

Create a new notebook for this task. Be sure to import the relevant libraries, along with your **`ords_prods_merge`**
dataframe, which should include your newly derived columns from the previous Exercise.

#### Import Data Set

In [2]:
path = r'C:\Users\facun\Desktop\Data Analysis\CF\PYTHON\Instacart Basket Analysis'

In [3]:
# Import df_ords_merged dataframe

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_prods_merged.pkl'))

In [4]:
df = ords_prods_merge

#### Data check

In [5]:
# Check shape to ensure nothing look odd abou the dimensions

df.shape

(32406041, 17)

In [6]:
# Double check the column names and first couple of rows

df.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   
5                   19.0         196                  1          1   
6                   20.0         196                  1          1   
7                   14.0         196                  2          1   
8                    0.0         196                  4          1   
9                   30.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   
5         Soda        77              7     9.0   both  Mid-range product   
6         Soda        77              7     9.0   both  Mid-range product   
7         Soda        77              7     9.0   both  Mid-range product   
8         Soda        77              7     9.0   both  Mid-range product   
9         Soda        77              7     9.0   both  Mid-range product   

     businest_day businest_period_of_day  
0  Regularly busy         Average orders  
1      Least busy         Average orders  
2      Least busy            Most orders  
3      Least busy         Average orders  
4      Least busy            Most orders  
5  Regularly busy         Average orders  
6    Busiest days            Most orders  
7    Busiest days            Most orders  
8    Busiest days            Most orders  
9      Least busy         Average orders

## Task 2

In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [7]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.213779
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.171095
12               15.887671
13               16.583536
14               16.771241
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Task 3

Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

In [8]:
# Create subset of the first million entries

df = ords_prods_merge[:1000000]

In [9]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

#### Answer:

The entire DataFrame provides a global view, while a subset analysis allows you to focus on specific subsets of the data. 

- The **Min** of the entire Dataframe Mean is **`15.213779`** while of the Subset is **`11.294069`** 

- The **Max** of the entire Dataframe Mean is **`22.902379`** while of the Subset is **`19.463012`** 

- Most of the **Mean** from the Subset Dataframe are greater than the Entire Dataframe with the expection of the **ID 17**



## Task 4

Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the **`transform()`** and **`loc()`** functions.

#### Create a loyalty flag using loc

 Create column "max_order"         =  dataframe grouped by "user_id" column  =    get the max of values(order_number) using transform function and **`np.max`** argument

In [10]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [11]:
pd.options.display.max_rows = None

In [12]:
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

In [13]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [14]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [15]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [16]:
# Check the frequency of your new “loyalty_flag” column:

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15877395
Loyal customer      10284396
New customer         6244250
Name: loyalty_flag, dtype: int64

## Task 5

The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [17]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'max', 'min']})

prices              
                       mean      max  min
loyalty_flag                             
Loyal customer    10.386229  99999.0  1.0
New customer      13.294406  99999.0  1.0
Regular customer  12.495501  99999.0  1.0

In [18]:
# Group the data by the 'loyalty_flag' column
loyalty_groups = ords_prods_merge.groupby('loyalty_flag')

# Calculate basic statistics for product prices within each group
loyal_customer_stats = loyalty_groups.get_group('Loyal customer')['prices'].describe()
regular_customer_stats = loyalty_groups.get_group('Regular customer')['prices'].describe()
new_customer_stats = loyalty_groups.get_group('New customer')['prices'].describe()

# Print the statistics for each loyalty category
print("Statistics for Loyal Customers:")
print(loyal_customer_stats)

print("\nStatistics for Regular Customers:")
print(regular_customer_stats)

print("\nStatistics for New Customers:")
print(new_customer_stats)


Statistics for Loyal Customers:
count    1.028440e+07
mean     1.038623e+01
std      3.280130e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.120000e+01
max      9.999900e+04
Name: prices, dtype: float64

Statistics for Regular Customers:
count    1.587740e+07
mean     1.249550e+01
std      5.397104e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

Statistics for New Customers:
count    6.244250e+06
mean     1.329441e+01
std      5.975479e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64


## Task 6

The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the `prices` of the items people are buying. `Create a `**`spending flag`** for each **user** based on the **average price** across all their orders using the following criteria:

- If the **mean** of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
- If the **mean** of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”


In [19]:
# Create new column "average price" using transform()

ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

#  Create column "average_price" 
#  dataframe grouped by "user_id" column  
#  get the mean of values(prices) using transform function and **`np.mean`** argument

In [20]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

     businest_day businest_period_of_day  max_order  loyalty_flag  \
0  Regularly busy         Average orders         10  New customer   
1      Least busy         Average orders         10  New customer   
2      Least busy            Most orders         10  New customer   
3      Least busy         Average orders         10  New customer   
4      Least busy            Most orders         10  New customer   

   average_price  
0       6.367797  
1       6.367797  
2       6.367797  
3       6.367797  
4       6.367797

In [21]:
# Rounding averages to make the number smaller

ords_prods_merge['average_price'] = ords_prods_merge['average_price'].round(2)

#### Create the flag "Low spender"

In [22]:
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low_spender'

#  If the average(mean) of the prices of the products by user is lower than 10 
# then(,) create a column "spending_flag where the customer will be labeled a `“Low_spender.”`

#### Create the flat "High spender"

In [23]:
ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High_spender'

#  If the average(mean) of the prices of the products by user is higher or equal to 10 
# then(,) create a column "spending_flag where the customer will be labeled a `“High_spender.”`

In [24]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...  department_id  prices  _merge    price_range_loc  \
0         Soda  ...              7     9.0    both  Mid-range product   
1         Soda  ...              7     9.0    both  Mid-range product   
2         Soda  ...              7     9.0    both  Mid-range product   
3         Soda  ...              7     9.0    both  Mid-range product   
4         Soda  ...              7     9.0    both  Mid-range product   

     businest_day businest_period_of_day max_order  loyalty_flag  \
0  Regularly busy         Average orders        10  New customer   
1      Least busy         Average orders        10  New customer   
2      Least busy            Most orders        10  New customer   
3      Least busy         Average orders        10  New customer   
4      Least busy            Most orders        10  New customer   

  average_price  spending_flag  
0          6.37    Low_spender  
1          6.37    Low_spender  
2          6.37    Low_spender  
3          6.37    Low_spender  
4          6.37    Low_spender  

[5 rows x 21 columns]

After executing the code to set the flags, print the frequency of your new **“spending_flag”** column using the **`value_counts()`** function:

In [25]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low_spender     31771140
High_spender      634901
Name: spending_flag, dtype: int64

## Task 7

In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine **frequent** versus **non-frequent** customers. Create an **order frequency flag** that marks the regularity of a **user’s ordering behavior** according to the **median** in the **“days_since_prior_order”** column. The criteria for the flag should be as follows:

- If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
- If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
- If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”


In [26]:
# Create new column "median_days" using transform()

ords_prods_merge['median_days_since_last_order'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

#  Create column "median_days" 
#  dataframe grouped by "user_id" column  
#  get the median of values(days_since_last_order) using transform function and **`np.median`** argument

In [27]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...  prices  _merge    price_range_loc    businest_day  \
0         Soda  ...     9.0    both  Mid-range product  Regularly busy   
1         Soda  ...     9.0    both  Mid-range product      Least busy   
2         Soda  ...     9.0    both  Mid-range product      Least busy   
3         Soda  ...     9.0    both  Mid-range product      Least busy   
4         Soda  ...     9.0    both  Mid-range product      Least busy   

  businest_period_of_day max_order  loyalty_flag  average_price spending_flag  \
0         Average orders        10  New customer           6.37   Low_spender   
1         Average orders        10  New customer           6.37   Low_spender   
2            Most orders        10  New customer           6.37   Low_spender   
3         Average orders        10  New customer           6.37   Low_spender   
4            Most orders        10  New customer           6.37   Low_spender   

   median_days_since_last_order  
0                          20.5  
1                          20.5  
2                          20.5  
3                          20.5  
4                          20.5  

[5 rows x 22 columns]

#### Create the flag "Non-frequent customer'

In [30]:
ords_prods_merge.loc[ords_prods_merge['median_days_since_last_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

#  If the median of the "days_prior_last_order" by user is higher than 20 
# then(,) create a column "order_frequency_flag" where the customer will be labeled a “Non-frequent customer.”

#### Create the flag "Regular customer"

In [33]:
ords_prods_merge.loc[(ords_prods_merge['median_days_since_last_order'] > 10) & (ords_prods_merge['median_days_since_last_order'] <=20), 'order_frequency_flag'] = 'Regular customer'

#  If the median of the "days_prior_last_order" by user is higher than 10 
# and lower or equal to 20 
# then(,) create a column "order_frequency_flag" where the customer will be labeled a “Regular customer.”

#### Create the flag "Frequent customer"

In [35]:
ords_prods_merge.loc[ords_prods_merge['median_days_since_last_order'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

#  If the median of the "days_prior_last_order" by user is lower or equal to 10 
# then(,) create a column "order_frequency_flag" where the customer will be labeled a “Frequent customer"

In [36]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...  _merge    price_range_loc    businest_day  \
0         Soda  ...    both  Mid-range product  Regularly busy   
1         Soda  ...    both  Mid-range product      Least busy   
2         Soda  ...    both  Mid-range product      Least busy   
3         Soda  ...    both  Mid-range product      Least busy   
4         Soda  ...    both  Mid-range product      Least busy   

  businest_period_of_day max_order  loyalty_flag average_price  spending_flag  \
0         Average orders        10  New customer          6.37    Low_spender   
1         Average orders        10  New customer          6.37    Low_spender   
2            Most orders        10  New customer          6.37    Low_spender   
3         Average orders        10  New customer          6.37    Low_spender   
4            Most orders        10  New customer          6.37    Low_spender   

  median_days_since_last_order   order_frequency_flag  
0                         20.5  Non-frequent customer  
1                         20.5  Non-frequent customer  
2                         20.5  Non-frequent customer  
3                         20.5  Non-frequent customer  
4                         20.5  Non-frequent customer  

[5 rows x 23 columns]

After executing the code to set the flags, print the frequency of your new **“order_frequency_flag”** column using the **`value_counts()`** function:

In [37]:
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21560324
Regular customer          7209005
Non-frequent customer     3636707
NaN                             5
Name: order_frequency_flag, dtype: int64

Check **null** values

In [41]:
ords_prods_merge[ords_prods_merge.order_frequency_flag.isnull()][['user_id', 'days_since_last_order', 'median_days_since_last_order', 'order_frequency_flag']]

user_id  days_since_last_order  median_days_since_last_order  \
13645692   159838                    NaN                           NaN   
17251990   159838                    NaN                           NaN   
17622767   159838                    NaN                           NaN   
24138593   159838                    NaN                           NaN   
25881080   159838                    NaN                           NaN   

         order_frequency_flag  
13645692                  NaN  
17251990                  NaN  
17622767                  NaN  
24138593                  NaN  
25881080                  NaN

Comment: I left the the values NaN as it might be a new customer

## Task 9

Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [42]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_prods_merge.pkl'))